### DISCUSSION TITLE
Error	LNK2019 unresolved external symbol __imp___strdup , already linked with WS2_32.LIB, GDI32.LIB, ADVAPI32.LIB, CRYPT32.LIB and USER32.LIB

### DISCUSSION THREAD

**Discussion by aniljadaun** [openssl/openssl#22797] [Q&A]
Error	LNK2019	unresolved external symbol __imp___strdup referenced in function _OPENSSL_INIT_set_config_appname
Error	LNK2001	unresolved external symbol __imp___strdup	
Error	LNK2019	unresolved external symbol __imp__fgets referenced in function _file_gets	
Error	LNK2001	unresolved external symbol __imp__fgets	
Error	LNK2019	unresolved external symbol __imp__strspn referenced in function _PEM_get_EVP_CIPHER_INFO	
Error	LNK2001	unresolved external symbol __imp__strspn
Error	LNK2001	unresolved external symbol __imp__strspn
Error	LNK1120	3 unresolved externals

**Comment by mspncp** [comment]
Sorry, but unless you provide more detailed information about your setup, we can only guess what causes your problem. 

For example

- what is your OpenSSL build configuration?
- does the error occur during the OpenSSL build process or when linking your application?
- what is the precise linker command?

The function `_strdup()` and its friends are implemented in the C runtime (CRT) and my guess is that you have some mixed CRT settings (/MT vs. /MD) in your project. A quick google search for [symbol __imp___strdup](https://www.google.com/search?q=+symbol+__imp___strdup) will bring you to pages like [this answer](https://learn.microsoft.com/en-us/answers/questions/576680/visual-studio-2017-c-linking-error-lnk2019-unresol) on learn.microsoft.com.

**Reply by aniljadaun** [reply]
HI @mspncp 
Thanks for your quick response . Is there a way to generate openssl libs with /MT flag ?

**Reply by mspncp** [reply]
If I understand it correctly, a static (resp. shared) OpenSSL library will use the static (resp. shared) CRT. In other words, this should be the case if you configured using `no-shared`.

https://github.com/openssl/openssl/blob/6d552a532754f6ee66d6cc604655deaeb5425b16/Configurations/10-main.conf#L1530-L1531

**Reply by mspncp** [reply]
Don't forget to do a clean rebuild after reconfiguring.

**Comment by mspncp** [comment]
BTW, you could have mentioned that you asked a related question previously in #22668.

**Comment by mspncp** [comment]
If this question is a duplicate of #22668 then according to @levitte's https://github.com/openssl/openssl/discussions/22668#discussioncomment-7519491 it should be no problem to link static openssl libraries and have shared CRT libraries.

**Reply by mspncp** [reply]
See here [Microsoft help](https://learn.microsoft.com/en-us/cpp/build/reference/zl-omit-default-library-name?view=msvc-170)

https://github.com/openssl/openssl/blob/6d552a532754f6ee66d6cc604655deaeb5425b16/Configurations/10-main.conf#L1518

**Reply by aniljadaun** [reply]
Hi @mspncp 

This is not duplicate. it is different issue. That issue was resolved using no-pinshared configure option.

Yes, i am trying to create no-shared i.e static libs with /MT compiler option.

 currently i am not sure how to do that. The error i shared initially is the error i am getting. We were previoulsy using openssl libs 1.0.2zg and we have two set of libs staticmt and staticmd. Currently MD project is successfully build but i am seeing linker issue if use same openssl 3.0 libs with MT project.
 
 Do i need to update this to use /MD option ?
  bin_cflags       => add(picker(debug   => "/MDd", 
                                release => sub { $disabled{shared} ? "/MT" : () },

**Reply by mspncp** [reply]
> That issue was resolved using no-pinshared configure option.

Discussion #22668 mentions none of this. If you managed to resolve the problem yourself, it would be nice if you could update the ticket accordingly and describe the solution as an answer to your own question for others to read.

**Reply by aniljadaun** [reply]
sure

**Reply by mspncp** [reply]
> Do i need to update this to use /MD option ?

Well, don't ask me, just try it ;-).  For trying, it should be sufficient to add the `/MD` switch explicitly to the configure call. Note also the comment about the `/MD` switch in release builds (the entire comment is larger than indicated here).

https://github.com/openssl/openssl/blob/6d552a532754f6ee66d6cc604655deaeb5425b16/Configurations/10-main.conf#L1527-L1531

**Reply by xqy-fly** [reply]
您好，请问问题解决了吗，我使用CMake打开NVIDIA Maxine AR SDK0.8.7 Release后，项目-属性-C/C++-代码生成-运行库-多线程DLL（/MD）修改成多线程（/MT）后报错：无法解析的外部符号 __imp__strdup，函数 _glfwAllocMonitor 中引用了该符号 我尝试了你所提供的所有方法，都没有解决问题
